### PR TITLE
mgr/dashboard: Set timeout in RestClient calls

### PR DESCRIPTION
**Pull request by votdev**
Signed-off-by: Volker Theile <vtheile@suse.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard/rest_client.py [hunk 1]
```diff
@@ -26,6 +27,16 @@
     from urllib3.exceptions import SSLError
 
 
+class TimeoutRequestsSession(requests.Session):
+    """
+    Set timeout argument for all requests if this is not already done.
+    """
+    def request(self, *args, **kwargs):
+        if kwargs.get('timeout') is None and Settings.REST_REQUESTS_TIMEOUT > 0:
```
Comment (LenzGr) [inline src/pybind/mgr/dashboard/rest_client.py]: Do we document these settings somewhere?

Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/rest_client.py]: This is IMO too generic. Is there any usage, where we explicitly change the timeout for a specific request? If not, why not simply call
```python
super(TimeoutRequestsSession, self).request(*args, **kwargs, timeout=Settings.REST_REQUESTS_TIMEOUT)
```
?

File: src/pybind/mgr/dashboard/rest_client.py [hunk 2]
```diff
@@ -465,6 +476,12 @@ def do_request(self,
             logger.exception("%s REST API failed %s: %s", self.client_name,
                              method.upper(), str(ex))
             raise RequestException(str(ex))
+        except Timeout as ex:
+            msg = "{} REST API {} timed out after {} seconds (url={}).".format(
+                self.client_name, ex.request.method, Settings.REST_REQUESTS_TIMEOUT,
+                ex.request.url)
+            logger.exception(msg)
+            raise RequestException(msg)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/rest_client.py]: Too bad, we cannot use exception chaining here. https://stackoverflow.com/a/792163/3185053

File: src/pybind/mgr/dashboard/rest_client.py [hunk 3]
```diff
@@ -465,6 +476,12 @@ def do_request(self,
             logger.exception("%s REST API failed %s: %s", self.client_name,
                              method.upper(), str(ex))
             raise RequestException(str(ex))
+        except Timeout as ex:
+            msg = "{} REST API {} timed out after {} seconds (url={}).".format(
+                self.client_name, ex.request.method, Settings.REST_REQUESTS_TIMEOUT,
+                ex.request.url)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/rest_client.py]: Can we make use of `str(ex)` here?

File: src/pybind/mgr/dashboard/tests/test_rest_client.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+import unittest
+
+from .. import mgr
+from ..rest_client import RestClient
+from mock import patch
+
+
+class RestClientTest(unittest.TestCase):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/tests/test_rest_client.py]: Another idea would be to actually execute a network request, similar to @p-na 's proxy test. And test the exception message. But I'm OK with this one.

File: src/pybind/mgr/dashboard/rest_client.py (hunk 3 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/rest_client.py]: The default error message does not provide all that information: 
``HTTPConnectionPool(host='dev-vtheile-ceph-1.oa.suse.de', port=8000): Read timed out. (read timeout=3)``

File: src/pybind/mgr/dashboard/tests/test_rest_client.py (hunk 4 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/tests/test_rest_client.py]: Wanted to make it as simple as possible.

**Comment by s0nea** [comment]
The last QA test run was not successful:

- https://shaman.ceph.com/builds/ceph/wip-tdehler-testing-23224/
- http://pulpito.ceph.com/tdehler-2018-07-30_10:46:54-rados:mgr-wip-tdehler-testing-23224-distro-basic-smithi/

**Comment by s0nea** [comment]
The last QA test run was successful:

- https://shaman.ceph.com/builds/ceph/wip-tdehler-testing-23224/74ee796c149a0316d79bbf1153ba333be9a1f6e4/
- http://pulpito.ceph.com/tdehler-2018-07-31_09:24:51-rados:mgr-wip-tdehler-testing-23224-distro-basic-smithi/

**Pull request merged**